src/main/scala/kamon/kamino/KaminoApiClient.scala

```diff
@@ -27,44 +33,81 @@ class KaminoApiClient(config: KaminoConfiguration) {
 
-    val request = () => {
+    val timeSinceLastPost = Duration.between(lastAttempt, clock.instant())
+    if(timeSinceLastPost.compareTo(config.clientBackoff) < 0) backoff
+
+    val request: () => Response = () => {
       val reqBody = RequestBody.create(MediaType.parse("application/octet-stream"), body)
       val request = new Request.Builder().url(apiUrl).post(reqBody).build
-      val response = client.newCall(request).execute
-      response.body().close()
-      response.code()
+      client.newCall(request).execute
     }
 
-    if(timeSinceLastPost.compareTo(config.clientBackoff) < 0) backoff(config.clientBackoff.toMillis)
-
-    def tryPosting: Try[Int] = Try {
+    def tryPosting: Try[Response] = Try {
       lastAttempt = clock.instant()
       request()
     }
 
-    tryPosting match {
-      case Success(200)  => // All good.
-      case _ if retries > 0 =>
-        backoff(config.retryBackoff.toMillis)
+    def parseResponse(response: Response): Try[IngestionResponse] = Try {
+      if(response.code() == 200) {
+        response.body().close()
+        IngestionResponse
+          .newBuilder()
+          .setStatus(OK)
+          .build()
+      } else {
+        IngestionResponse.parseFrom(response.body().bytes())
+      }
+    }
+
+    tryPosting.flatMap(parseResponse) match {
+      case Success(ingestionResult) =>
+        ingestionResult.getStatus match {
+          case OK =>
+            logger.info("Snapshot ingested")
+          case STALE =>
+            logger.warn("Ingestion declined, stale data")
+          case BLOCKED =>
+            logger.warn("Ingestion declined, plan limits reached")
+          case UNAUTHORIZED =>
+            logger.error("Ingestion declined, missing or wrong API key")
+          case CORRUPTED =>
+            logger.warn("Ingestion declined, illegal batch")
+          case ERROR =>
+            logger.warn("Ingestion declined, unknown error")
+            backoff
+            postWithRetry(body, apiUrl, retries - 1)
+        }
+      case Failure(connectionException) if retries > 0 =>
+        logger.error(s"Connection error, retrying... ($retries left) ${connectionException.getMessage}")
+        backoff
         postWithRetry(body, apiUrl, retries - 1)
-      case _ => // Nothing to do.
+      case Failure(connectionException) =>
+        logger.error(s"Ingestion error, no retries, dropping snapshot... ${connectionException.getMessage}")
+
     }
+
   }
 
   private def createHttpClient(config: KaminoConfiguration): OkHttpClient = {
-    new OkHttpClient.Builder()
+    val builder = new OkHttpClient.Builder()
       .connectTimeout(config.connectionTimeout.toMillis, TimeUnit.MILLISECONDS)
       .readTimeout(config.readTimeout.toMillis, TimeUnit.MILLISECONDS)
-      .build()
-  }
 
+    config.proxy.foreach { pt =>
+      builder.proxy(
+        new Proxy(pt, new InetSocketAddress(config.proxyHost, config.proxyPort))
```

Comment:
just out of curiosity, proxies are added the same way regarding of them being SOCKS or HTTPS?

Reply:
Configured type is mapped to `pt: Proxy.Type` enum and passed to the constructor.

Reply:
I meant in the OkHttp builder